src/pydvl/influence/torch/util.py

```diff
@@ -598,3 +608,141 @@ def __init__(self, original_exception: RuntimeError):
             f" Inspect the original exception message: \n{str(original_exception)}"
         )
         super().__init__(err_msg)
+
+
+LossType = Callable[[torch.Tensor, torch.Tensor], torch.Tensor]
+
+
+class BlockMode(Enum):
+    """
+    Enumeration for different modes of grouping model parameters.
+
+    Attributes:
+        LAYER_WISE: Groups parameters by layers of the model.
+        PARAMETER_WISE: Groups parameters individually.
+        FULL: Groups all parameters together.
+    """
+
+    LAYER_WISE: str = "layer_wise"
+    PARAMETER_WISE: str = "parameter_wise"
+    FULL: str = "full"
+
+
+@dataclass
+class ModelParameterDictBuilder:
+    """
+    A builder class for creating ordered dictionaries of model parameters based on
+    specified block modes or custom blocking structures.
+
+    Attributes:
+        model: The neural network model.
+        detach: Whether to detach the parameters from the computation graph.
+    """
+
+    model: torch.nn.Module
+    detach: bool = True
+
+    def _optional_detach(self, p: torch.nn.Parameter):
+        if self.detach:
+            return p.detach()
+        return p
+
+    def build(self, block_structure: OrderedDict[str, List[str]]):
+        """
+        Builds an ordered dictionary of model parameters based on the specified block
+        structure represented by an ordered dictionary, where the keys are block
+        identifiers and the values are lists of model parameter names contained in
+        this block.
+
+        Args:
+            block_structure: The block structure specifying how to group the parameters.
+
+        Returns:
+            An ordered dictionary of ordered dictionaries, where the outer dictionary's
+            keys are block identifiers and the inner dictionaries map parameter names
+            to parameters.
+        """
+        parameter_dict = {}
+
+        for block_name, parameter_names in block_structure.items():
+            inner_ordered_dict = {}
+            for parameter_name in parameter_names:
+                parameter = self.model.state_dict()[parameter_name]
+                if parameter.requires_grad:
+                    inner_ordered_dict[parameter_name] = self._optional_detach(
+                        parameter
+                    )
+                else:
+                    warnings.warn(
+                        f"The parameter {parameter_name} from the block "
+                        f"{block_name} is mark as not trainable in the model "
+                        f"and will be excluded from the computation."
+                    )
+            parameter_dict[block_name] = inner_ordered_dict
+
+        return parameter_dict
+
+    def build_from_block_mode(
```

Comment:
I would have expected that `build_from_block_mode` creates a `block_structure` and then calls `build`. Are there specific reasons why would not work here?

Reply:
Will check, if this can be simplified and made more readable.

Reply:
Done